src/input_popup.cpp

```diff
@@ -0,0 +1,381 @@
+        if( description_monofont ) {
+            cataimgui::PushMonoFont();
+        }
+        // todo: you should not have to manually split text to print it as paragraphs
+        uint64_t offset = 0;
+        uint64_t pos = 0;
+        while( pos != std::string::npos ) {
+            pos = description.find( "\n", offset );
+            std::string str = description.substr( offset, pos - offset );
+            cataimgui::TextColoredParagraph( description_default_color, str, std::nullopt,
+                                             str_width_to_pixels( width ) );
+            ImGui::NewLine();
+            offset = pos + 1;
+        }
+        if( description_monofont ) {
+            ImGui::PopFont();
+        }
+    }
+
+    if( !label.empty() ) {
+        ImGui::AlignTextToFramePadding();
+        cataimgui::draw_colored_text( label, c_light_red );
+        ImGui::SameLine();
+    }
+
+    // make sure the cursor is in the input field when we regain focus
+    if( !ImGui::IsWindowFocused() ) {
+        set_focus = true;
+    }
+
+    if( set_focus && ImGui::IsWindowFocused() ) {
+        ImGui::SetKeyboardFocusHere();
+        set_focus = false;
+    }
+    draw_input_control();
+}
+
+bool input_popup::handle_custom_callbacks( const std::string &action )
+{
+
+    input_event ev = ctxt.get_raw_input();
+    int key = ev.get_first_input();
+
+    bool next_loop = false;
+    for( const auto &cb : callbacks ) {
+        if( cb.first == callback_input( action, key ) && cb.second() ) {
+            next_loop = true;
+            break;
+        }
+    }
+
+    return next_loop;
+}
+
+string_input_popup_imgui::string_input_popup_imgui( int width, const std::string &old_input,
+        const std::string &title, const point &pos, ImGuiWindowFlags flags ) :
+    input_popup( width, title, pos, flags ),
+    old_input( old_input )
+{
+    ctxt.register_action( "TEXT.UP" );
+    ctxt.register_action( "TEXT.DOWN" );
+
+    set_text( old_input );
+}
+
+static int history_callback( ImGuiInputTextCallbackData *data )
+{
+    string_input_popup_imgui *popup = static_cast<string_input_popup_imgui *>( data->UserData );
+
+    if( data->EventFlag == ImGuiInputTextFlags_CallbackHistory ) {
+        popup->update_input_history( data );
+    }
+
+    return 0;
+}
+
+void string_input_popup_imgui::draw_input_control()
+{
+    ImGuiInputTextCallback callback = nullptr;
+    void *data = nullptr;
+    ImGuiInputTextFlags flags = ImGuiInputTextFlags_None;
+
+    if( !is_uilist_history ) {
+        callback = history_callback;
+        data = this;
+        flags |= ImGuiInputTextFlags_CallbackHistory;
+    }
+
+    // shrink width of input field if we only allow short inputs
+    float input_width = str_width_to_pixels( max_input_length );
+    if( input_width < ImGui::CalcItemWidth() ) {
+        ImGui::SetNextItemWidth( input_width );
+    }
+
+    ImGui::InputText( "##string_input", text.data(), max_input_length, flags, callback, data );
```

Comment:
ImGui comes with a helper function that lets you directly use a `std::string` as an argument here. You’ll have to import the code [from the ImGui repository](https://github.com/ocornut/imgui/blob/master/misc/cpp/imgui_stdlib.cpp), but it does string resizing properly.

Reply:
Yes, I've seen that and also tried it out and it seemed to work properly, but @katemonster33 pointed out that casting `std::string::c_str` to non-const to be written to is undefined behavior, so I refrained from adding it.

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/c_str is what I was referencing when I called it undefined behavior, but data() points to the same piece of memory (or it should) and the only difference is the non-const, it also doesn't have the note about undefined behavior attached to it. If we prefer the imgui_stdlib method then I'm fine with it

Reply:
Made #77979 for this.